src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,186 @@
+            }
+            else
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Your image has been uploaded. %1 left. Please "
+                            "wait until all of them are uploaded. About %2 "
+                            "seconds left.")
+                        .arg(uploadQueue.size())
+                        .arg(uploadQueue.size() * 3)));
+                // Argument number 2 is the ETA.
+                // 2 seconds for the timer that's there not to spam Nuuls' server
+                // and 1 second of actual uploading.
+            }
+            isUploading = false;
+            if (uploadQueue.size())
+            {
+                QTimer::singleShot(2000, [channel, &textEdit]() {
+                    uploadImageToNuuls(uploadQueue.front(), channel, textEdit);
+                    uploadQueue.pop();
+                });
+            }
+            return Success;
+        })
+        .onError([channel](NetworkResult result) -> bool {
+            channel->addMessage(makeSystemMessage(
+                QString("An error happened while uploading your image: %1")
+                    .arg(result.status())));
+            isUploading = false;
+            return true;
+        })
+        .execute();
+}
+
+void upload(const QMimeData *source, ChannelPtr channel,
+            ResizingTextEdit &outputTextEdit)
+{
+    if (isUploading)
+    {
+        channel->addMessage(makeSystemMessage(
+            QString("Please wait until the upload finishes.")));
+        return;
+    }
+
+    isUploading = true;
+    channel->addMessage(makeSystemMessage(QString("Started upload...")));
+
+    if (source->hasFormat("image/png"))
+    {
+        uploadImageToNuuls({source->data("image/png"), "png"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/jpeg"))
+    {
+        uploadImageToNuuls({source->data("image/jpeg"), "jpeg"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/gif"))
+    {
+        uploadImageToNuuls({source->data("image/gif"), "gif"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasUrls())
+    {
+        for (const QUrl &path : source->urls())
+        {
+            if (!getImageFileFormat(path.toLocalFile()).isEmpty())
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading image: %1").arg(path.toLocalFile())));
+                QImage img = QImage(path.toLocalFile());
+                if (img.isNull())
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Couldn't load image :(")));
+                    return;
+                }
+                QByteArray imageData;
+                QBuffer buf(&imageData);
+                buf.open(QIODevice::WriteOnly);
+                img.save(&buf, "png");
```

Comment:
we're discarding the *success* state here, that seems bad when we're trying to make a conversion